extensions/amp-fx-collection/0.1/scroll-toggle.js

```diff
@@ -185,24 +219,132 @@ export function assertValidScrollToggleElement(element, computedStyle) {
 export function getScrollTogglePosition(element, type, computedStyle) {
   const position = type.replace(/^float\-in\-([^\s]+)$/, '$1');
 
-  devAssert(position.length > 0);
+  devAssert(position == 'top' || position == 'bottom');
 
   // naming convention win:
   // position `top` should have `top: 0` and `bottom` should have `bottom: 0`
   if (!assertStyleOrWarn(
-      computedStyle, position, px(0), element, `amp-fx=${type}`)) {
+      computedStyle, position, px(0), element, withAmpFxType(type))) {
     return null;
   }
 
   return /** @type {!ScrollTogglePosition} */ (position);
 }
 
+
+/**
+ * @param {!Element} element
+ * @param {boolean} isShown
+ * @param {!ScrollTogglePosition} position
+ */
+function scrollToggle(element, isShown, position) {
+  let offset = 0;
+  Services.resourcesForDoc(element).measureMutateElement(element, () => {
+    offset = getScrollToggleFloatInOffset(element, isShown, position);
+  }, () => {
+    scrollToggleFloatIn(element, offset);
+  });
+}
+
 /**
  * MUST be done inside mutate phase.
+ * @param {!ScrollToggleDispatch} dispatch
  * @param {!Element} element
+ * @param {!ScrollTogglePosition} position
  */
-export function installScrollToggleStyles(element) {
-  setImportantStyles(element, {'will-change': 'transform'});
+export function installScrollToggleFloatIn(dispatch, element, position) {
+  const viewport = Services.viewportForDoc(element);
+
+  if (viewport.getPaddingTop() <= 0) {
+    dispatch.observe(isShown => {
+      scrollToggle(element, isShown, position);
+    });
+  }
+
+  let lastEvent = 0;
+
+  // Use viewport's "padding top" as a proxy signal for whether there's a
+  // viewer header displayed. If displayed, we display float-in-* elements as
+  // well.
+  viewport.setFixedElementMeasurer(element,
+      (afterAnimation, prevPaddingTop, paddingTop) => {
+        const curEvent = ++lastEvent;
+
+        const resources = Services.resourcesForDoc(element);
+        const isShown = isShownPerViewportPaddingTop(paddingTop);
+
+        // Disable scroll dispatch to rely on viewport instead.
+        if (isShown) {
+          dispatch.dispose();
+        }
+
+        (isShown ? Promise.resolve() : afterAnimation).then(() => {
+          resources.mutateElement(element, () => {
+            if (lastEvent != curEvent) {
+              return;
+            }
+            setVisibilityStyles(element, isShown);
+          });
+        });
+
+        const {top, bottom, animOffset} = measureFloatInFromViewport(
+            element, position, prevPaddingTop, paddingTop);
+
+        resources.mutateElement(element, () => {
+          if (top !== undefined) {
+            setStyle(element, 'top', px(top));
```

Comment:
Why does the element need to keep it's in FixedLayer if you're just gonna set it here?

Reply:
Not sure I follow the question? The element stays *transferred*, but is not tracked by `FixedLayer`. I feel like I'm gonna go 🍌🍌🍌 if I have to explain why it needs to stay transferred again :)

Reply:
The `keepOffsetTop` parameter. We're just gonna override it? Why keep the offset at all?

Reply:
We override _when the padding top changes_. Keeping it prevents it from flashing after removal.